Add new ELMoTransformerEmbeddings class

stefan-it opened the pull request:
Hi,

this PR introduces a new `ELMoTransformerEmbeddings` class. With the [help](https://github.com/allenai/allennlp/issues/2351) from @brendan-ai2 it is possible to get embeddings from a transformer-based ELMo model.

That model was proposed in [Dissecting Contextual Word Embeddings: Architecture and Representation](http://aclweb.org/anthology/D18-1179).

Embeddings from a transformed-based ELMo model can now be used in `flair`. New CUDA semantic is also used. Training a model both on CPU and GPU works.

A pretrained transformer-based ELMo model for Basque can be downloaded from:

```bash
wget https://schweter.eu/cloud/elmo-transformer-models/eu-elmo-transformer-model.tar.gz
```

---

### Downstream task example

To train a model for PoS tagging (in this example for Basque) with the new `ELMoTransformerEmbeddings` just follow the following instructions:

Clone a recent version of `allennlp` and install it, e.g.:

```bash
git clone https://github.com/allenai/allennlp.git
cd allennlp
pip3 install -e .
```

I tested it with commit 4c5de57 of `allennlp`.

Then download a pretrained transformer-based ELMo model.

```bash
wget https://schweter.eu/cloud/elmo-transformer-models/eu-elmo-transformer-model.tar.gz
``` 

The training can be started with the following script:

```python
from typing import List

from flair.data_fetcher import NLPTaskDataFetcher, NLPTask
from flair.embeddings import TokenEmbeddings, WordEmbeddings, StackedEmbeddings, ELMoTransformerEmbeddings

corpus = NLPTaskDataFetcher.load_corpus(NLPTask.UD_BASQUE).downsample(0.1)

tag_type = 'upos'
tag_dictionary = corpus.make_tag_dictionary(tag_type=tag_type)

custom_embedding = WordEmbeddings('eu')

embedding_types: List[TokenEmbeddings] = [
    custom_embedding,
    ELMoTransformerEmbeddings(model_file='eu-elmo-transformer-model.tar.gz')
]

embeddings: StackedEmbeddings = StackedEmbeddings(embeddings=embedding_types)

from flair.models import SequenceTagger

tagger: SequenceTagger = SequenceTagger(hidden_size=512,
                                        embeddings=embeddings,
                                        tag_dictionary=tag_dictionary,
                                        tag_type=tag_type,
                                        use_crf=False,
                                        pickle_module='dill')

from flair.trainers import ModelTrainer
from flair.training_utils import EvaluationMetric

trainer: ModelTrainer = ModelTrainer(tagger, corpus)

trainer.train(f'resources/taggers/ud-basque-elmo-transformer',
              EvaluationMetric.MICRO_ACCURACY,
              learning_rate=0.1,
              mini_batch_size=8,
              max_epochs=1)
```

tabergma commented on flair/embeddings.py:
```diff
@@ -326,6 +326,84 @@ def extra_repr(self):
+            for token, token_idx in zip(sentence.tokens, range(len(sentence.tokens))):
+                token: Token = token
+                embedding = embeddings[token_idx]
+                word_embedding = torch.autograd.Variable(torch.FloatTensor(embedding))
```
`Tensor` and `Variable` was merged into one class with PyTorch 0.4 (see https://pytorch.org/blog/pytorch-0_4_0-migration-guide/#merging-tensor-and-variable-and-classes). 
So can't you just do
```
word_embedding = torch.FloatTensor(embedding)
```

stefan-it replied:
Thanks :) I fixed that (and it works without problems both on CPU and GPU)!

alanakbik commented:
I am getting the following error when running the script: 

```console
Traceback (most recent call last):
  File "/home/aakbik/PycharmProjects/flair/local_test_local.py", line 15, in <module>
    ELMoTransformerEmbeddings(model_file='/home/aakbik/Documents/Data/Embeddings/eu-elmo-transformer-model.tar.gz')
  File "/home/aakbik/PycharmProjects/flair/flair/embeddings.py", line 337, in __init__
    self.allen_nlp_utils = ELMoTransformerEmbeddings.AllenNlpUtils(model_file)
  File "/home/aakbik/PycharmProjects/flair/flair/embeddings.py", line 360, in __init__
    requires_grad=False
  File "/home/aakbik/.environments/flair/lib/python3.6/site-packages/allennlp/modules/token_embedders/bidirectional_language_model_token_embedder.py", line 68, in __init__
    archive = load_archive(archive_file, overrides=json.dumps(overrides))
  File "/home/aakbik/.environments/flair/lib/python3.6/site-packages/allennlp/models/archival.py", line 156, in load_archive
    cuda_device=cuda_device)
  File "/home/aakbik/.environments/flair/lib/python3.6/site-packages/allennlp/models/model.py", line 321, in load
    return cls.by_name(model_type)._load(config, serialization_dir, weights_file, cuda_device)
  File "/home/aakbik/.environments/flair/lib/python3.6/site-packages/allennlp/common/registrable.py", line 58, in by_name
    raise ConfigurationError("%s is not a registered name for %s" % (name, cls.__name__))
allennlp.common.checks.ConfigurationError: 'language_model is not a registered name for Model'
```

Since I haven't worked with allennlp much: Any ideas where this error is coming from?

alanakbik commented on flair/embeddings.py:
```diff
@@ -326,6 +326,84 @@ def extra_repr(self):
+            self.lm_embedder = BidirectionalLanguageModelTokenEmbedder(
+                archive_file=model_file,
+                dropout=0.2,
+                bos_eos_tokens=["<S>", "</S>"],
```
should be `bos_eos_tokens=("<S>", "</S>"),` instead. `bos_eos_tokens` expects Tuple, not List.

stefan-it replied:
Thanks, I've fix it in 102095f :)

stefan-it commented:
@alanakbik You should use a recent `master` version from `allennlp` -> I think the one from pip is too old.

alanakbik commented:
Ah ok, I will!

stefan-it commented:
@alanakbik It would work with the latest `allennlp` version, but for saving the trained model the `dill` package needs to be used:

```python
torch.save(model_state, str(model_file), pickle_module=dill)
```

Do you think we can add a new parameter to `SequenceTagger` that allows to use another library for saving a torch model?

alanakbik commented:
@stefan-it yes I think that would be better since it would make it easier for people to use the class (just install allennlp and dill)!

stefan-it commented:
I rebased the code on latest `master` and refactored the `ELMoTransformerEmbeddings` class code. I also introduced a new helper method for saving a torch model using an user-defined pickle module (like `dill`, but default is the normal `pickle` module).

tabergma commented:
mmhh.. I guess something went wrong with the rebasing as the changes which are already in the master branch are shown as changes of this PR. Would you mind updating the branch again so that we only see the changes you actually did? Thanks!

stefan-it commented:
Sorry for the confusion!

alanakbik commented:
Hi @stefan-it - I'm testing the current version locally with allennlp 8.1. When running this code:

```python
embeddings = ELMoTransformerEmbeddings('eu-elmo-transformer-model.tar.gz')
embeddings.embed(Sentence('I love Berlin'))
```

I get the error: 

```console
Traceback (most recent call last):
  File "/home/aakbik/PycharmProjects/stefan-flair/flair/train.py", line 21, in <module>
    embeddings = ELMoTransformerEmbeddings('eu-elmo-transformer-model.tar.gz')
  File "/home/aakbik/PycharmProjects/stefan-flair/flair/flair/embeddings.py", line 356, in __init__
    requires_grad=False
  File "/home/aakbik/.environments/stefan/lib/python3.6/site-packages/allennlp/modules/token_embedders/bidirectional_language_model_token_embedder.py", line 68, in __init__
    archive = load_archive(archive_file, overrides=json.dumps(overrides))
  File "/home/aakbik/.environments/stefan/lib/python3.6/site-packages/allennlp/models/archival.py", line 156, in load_archive
    cuda_device=cuda_device)
  File "/home/aakbik/.environments/stefan/lib/python3.6/site-packages/allennlp/models/model.py", line 321, in load
    return cls.by_name(model_type)._load(config, serialization_dir, weights_file, cuda_device)
  File "/home/aakbik/.environments/stefan/lib/python3.6/site-packages/allennlp/common/registrable.py", line 58, in by_name
    raise ConfigurationError("%s is not a registered name for %s" % (name, cls.__name__))
allennlp.common.checks.ConfigurationError: 'language_model is not a registered name for Model'
```

Any idea where this error comes from?

stefan-it commented:
*0.8.1* is too old. Please try the latest `master` 2b5249 - I also tested 55b9b.

With the latest `master` version you should see something like:

```python
from flair.data import Sentence
from flair.embeddings import ELMoTransformerEmbeddings

sentence = Sentence("It is no longer snowing in Munich .")
embeddings = ELMoTransformerEmbeddings(model_file='eu-elmo-transformer-model.tar.gz')
embeddings.embed(sentence)

for token in sentence.tokens:
    print(token.embedding)
```

```bash
tensor([  6.3030, -25.0204,   3.3425,  ...,   4.0137, -29.3740,   7.8902])
tensor([ 12.4846, -19.3540,  -3.5304,  ...,   6.7518,  -6.8744,  -7.1798])
tensor([ 16.5951, -19.1923,  -6.8334,  ...,  12.1934,  -0.0000,  -0.0000])
tensor([ 4.6706, -0.0000, -0.0000,  ...,  6.8242, -8.0350, -3.6316])
tensor([  5.2432, -15.5042,  -4.4524,  ...,   2.1873, -12.7269,  -5.6938])
tensor([ 20.1224, -21.2971,   2.6443,  ...,   9.2506,   1.6562,  -4.3060])
tensor([ 2.4384, -0.0000, -7.1844,  ...,  0.0000, -0.0000,  3.2789])
tensor([ 16.3643,  -0.0000,  -0.0000,  ..., -14.8744,  16.0890,  13.4172])
```

alanakbik commented:
Thanks this works - sorry with the dill change I somehow thought this meant that it works with pip install allennlp :)

For now, we could include this class but advise people that it is experimental and requires to check out the current master branch of allennlp. As soon as a new version of allennlp is pushed to pip that allows this class to work we could remove the experimental tag. What do you think?

stefan-it commented:
I fully agree with you :) Even in the `allennlp` library the ELMo Transformer is experimental, see their disclaimer:

```
ExperimentalFeatureWarning: This particular transformer implementation is a provisional feature that's intended for AI2 internal use and might be deleted at any time
```

You can add me as a kind of maintainer for the `ELMoTransformerEmbeddings` class, because I'll use it regulary :)

alanakbik commented:
Ok, sounds good! Will merge as soon as tests run through.
